### PR TITLE
packages/core-api,test-utils

### PR DESCRIPTION
**Pull request by Rugvip**
Adding some more building blocks to make it easier to test components inside an app with APIs.

This makes the ApiProvider forward requests to parent contexts if available. I figured that's a nice way to override APIs inside tests without having to provide direct access to the test registry. It's also likely going to be used to wrap plugins with plugin-specific APIs as well.

It also makes it a bit cleaner to construct test registries by adding `ApiRegistry.with(ref, impl)`, which can also be changed to add multiple APIs. We may want to get rid of the `ApiRegistry.from` altogether tbh.

Also adding `renderInTestApp`, which does the render + wrapInTestApp, but also uses the async render with effects to avoid act warnings everywhere, which we'd be getting otherwise.

The idea for testing with API mocks is to provide a set of default mocks for all core APIs that have some sane default behaviour for most tests. Then if a test needs to dig a bit deeper and assert how the APIs are consumed, you'd override an API with the following pattern:

```tsx
const mockAlertApi = new MockAlertApi();

const rendered = await renderInTestApp(
  <ApiProvider apis={ApiRegistry.with(alertApiRef, mockAlertApi)}>
    <MyComponent />
  </ApiProvider>,
);

expect(mockAlertApi.getAlerts()).toEqual([{ message: 'my-alert' }]);
```

Some other approaches considered:

```tsx
const mockAlertApi = new MockAlertApi();

const rendered = await renderInTestApp(<MyComponent />, {
  apis: [
    {
      implements: alertApiRef,
      factory: () => mockAlertApi,
    },
  ],
});

expect(mockAlertApi.getAlerts()).toEqual([{ message: 'my-alert' }]);
```

```tsx
const mockAlertApi = new MockAlertApi();

getTestRegistry().register(alertApiRef, mockAlertApi);
const rendered = await renderInTestApp(<MyComponent />);

expect(mockAlertApi.getAlerts()).toEqual([{ message: 'my-alert' }]);
```

I like the absence of magic in the first alternative though.

Anyway, all of that stuff ^ coming in another PR, but feel free to leave comments here 😁

### CODE REVIEW

File: packages/core-api/src/apis/ApiProvider.test.tsx [hunk 1]
```diff
@@ -51,6 +51,35 @@ describe('ApiProvider', () => {
...
+          [aRef, 'x'],
+          [bRef, 'y'],
+        ])}
+      >
+        <ApiProvider apis={ApiRegistry.from([[aRef, 'z']])}>
+          <MyComponent />
+        </ApiProvider>
+      </ApiProvider>,
+    );
+    renderedHook.getByText('a=z b=y');
```
Comment (benjdlambert) [inline packages/core-api/src/apis/ApiProvider.test.tsx:80]: I know it's not needed, but is it still good practice to have an assert here?

File: packages/test-utils/src/testUtils/appWrappers.test.tsx [hunk 2]
```diff
@@ -14,22 +14,57 @@
...
+        wrapInTestApp(
+          <>
+            <Route path="/route1">Route 1</Route>
+            <Route path="/route2">Route 2</Route>
+          </>,
+          { routeEntries: ['/route2'] },
+        ),
+      );
+      expect(rendered.getByText('Route 2')).toBeInTheDocument();
+      await Promise.resolve();
```
Comment (benjdlambert) [inline packages/test-utils/src/testUtils/appWrappers.test.tsx:36]: Why is this needed?

File: packages/core-api/src/apis/ApiProvider.test.tsx (hunk 1 above)
Comment (Rugvip) [reply packages/core-api/src/apis/ApiProvider.test.tsx:80]: I do it when I haven't yet disabled the lint rule and can't be arsed xP

I dunno, what'd be the benefit of doing it? So far I haven't hit any trouble because of omitting it, but could be convinced :grin:

Comment (benjdlambert) [reply packages/core-api/src/apis/ApiProvider.test.tsx:80]: I mean I could be convinced that it's not a good idea too.. haha.

I just think it reads a little strange. but so does:

```ts
expect(() => renderedHook.getByText(..)).not.toThrow()
```

Maybe something like:

```ts
expect(await renderedHook.getByText('...')).toBeInTheDocument()
```

is nicer?

Comment (benjdlambert) [reply packages/core-api/src/apis/ApiProvider.test.tsx:80]: or omit the await if it's not needed

Comment (benjdlambert) [reply packages/core-api/src/apis/ApiProvider.test.tsx:80]: just conistency I guess, as other tests to the latter.

File: packages/test-utils/src/testUtils/appWrappers.test.tsx (hunk 2 above)
Comment (Rugvip) [reply packages/test-utils/src/testUtils/appWrappers.test.tsx:38]: It's waiting for the warnings to be generated, added a comment

File: packages/core-api/src/apis/ApiProvider.test.tsx (hunk 1 above)
Comment (Rugvip) [reply packages/core-api/src/apis/ApiProvider.test.tsx:80]: yeah `expect(renderedHook.getByText('...')).toBeInTheDocument()` would be the alternative, but still cba using that everywhere :p

**Review by benjdlambert — APPROVED** [review]
